File: projects/timely.noizu.com/apps/android/app/src/main/java/com/noizu/timely/data/local/Entities.kt
package com.noizu.timely.data.local

import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * The local mirror.
 *
 * Room is the UI's only source of truth. Screens never read from the network;
 * they observe these tables, and the sync engine is the only writer of
 * server-derived state. That is what makes the app fully usable offline --
 * there is no code path where a stale token can make a screen go blank.
 *
 * Every table carries the sync envelope columns (`server_revision`,
 * `deleted_at`, `origin_device_id`) because the conflict rules in section 8.2
 * are expressed in terms of them. `deleted_at` is a tombstone: rows are never
 * hard-deleted by sync, only marked, so that a delete replayed after a local
 * edit still resolves deterministically.
 *
 * `pending_*` columns hold local, not-yet-acknowledged edits. A row can be both
 * server-derived and locally dirty at once; the UI shows the local value and a
 * "not yet synced" affordance.
 */

@Entity(
    tableName = "clients",
    indices = [Index(value = ["workspaceId", "canonicalName"], unique = true)],
)
data class ClientEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val name: String,
    val canonicalName: String,
    val notes: String = "",
    val autoCreated: Boolean = false,
    val mergedIntoId: String? = null,
    val reviewState: String = "unreviewed",
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
    /** True while a local create/edit for this row is still in the push queue. */
    val pendingLocal: Boolean = false,
)

@Entity(
    tableName = "projects",
    indices = [
        Index(value = ["workspaceId", "clientId", "canonicalName"], unique = true),
        Index(value = ["clientId"]),
    ],
)
data class ProjectEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val name: String,
    val canonicalName: String,
    val clientId: String? = null,
    val clientName: String = "",
    val notes: String = "",
    val autoCreated: Boolean = false,
    val mergedIntoId: String? = null,
    val reviewState: String = "unreviewed",
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
    val pendingLocal: Boolean = false,
)

@Entity(
    tableName = "tickets",
    indices = [
        Index(value = ["workspaceId", "projectId", "canonicalName"], unique = true),
        Index(value = ["projectId"]),
    ],
)
data class TicketEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val name: String,
    val canonicalName: String,
    val clientId: String? = null,
    val projectId: String? = null,
    val clientName: String = "",
    val projectName: String = "",
    val notes: String = "",
    val autoCreated: Boolean = false,
    val mergedIntoId: String? = null,
    val reviewState: String = "unreviewed",
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
    val pendingLocal: Boolean = false,
)

@Entity(
    tableName = "time_spans",
    indices = [
        Index(value = ["workspaceId", "startEpochSeconds"]),
        Index(value = ["projectId"]),
        Index(value = ["reviewState"]),
    ],
)
data class TimeSpanEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val title: String = "",
    val clientId: String? = null,
    val projectId: String? = null,
    val ticketId: String? = null,
    val clientName: String = "",
    val projectName: String = "",
    val ticketName: String = "",
    val start: String,
    val end: String? = null,
    /**
     * Denormalized epoch seconds for the two things the timeline actually does:
     * order by time, and filter to a day. Sorting on the ISO string would be
     * correct only for `Z`-suffixed instants of identical precision, which is
     * exactly the assumption a server upgrade quietly breaks.
     */
    val startEpochSeconds: Long,
    val endEpochSeconds: Long? = null,
    val source: String = "manual",
    val isBillable: Boolean = false,
    val notes: String = "",
    val reviewState: String = "unreviewed",
    /** `ReviewReasonDto` list, stored as the raw JSON array. */
    val reviewReasonsJson: String = "[]",
    val derivedFromSpanIdsJson: String = "[]",
    val lockedAt: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val updatedAtEffective: String? = null,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
    val pendingLocal: Boolean = false,
)

@Entity(
    tableName = "screenshots",
    indices = [Index(value = ["spanId"]), Index(value = ["workspaceId", "capturedAt"])],
)
data class ScreenshotEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val spanId: String? = null,
    val capturedAt: String,
    val fileName: String = "",
    val activeAppName: String = "",
    val uploadState: String = "local_only",
    val blobAvailable: Boolean = false,
    val blobContentHash: String? = null,
    val blobByteSize: Long? = null,
    val blobUploadedAt: String? = null,
    val blobUrl: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
)

@Entity(tableName = "vision_analyses", indices = [Index(value = ["screenshotId"])])
data class VisionAnalysisEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val screenshotId: String,
    val analyzedAt: String,
    val model: String = "",
    val statusUpdate: String = "",
    val inferredProject: String = "",
    val inferredTask: String = "",
    val projectSwitchDetected: Boolean = false,
    val confidence: Double = 0.0,
    val evidence: String = "",
    val privacySensitive: Boolean = false,
    val privacyCategory: String = "none",
    val rawResponseWithheld: Boolean = true,
    val errorMessage: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
)

@Entity(tableName = "censored_screenshots", indices = [Index(value = ["screenshotId"])])
data class CensoredScreenshotEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val screenshotId: String,
    val spanId: String? = null,
    val fileName: String = "",
    val activeAppName: String = "",
    val capturedAt: String,
    val censoredAt: String,
    val model: String = "",
    val category: String = "none",
    val reason: String = "",
    val confidence: Double = 0.0,
    val deletedLocalFile: Boolean = false,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
)

@Entity(tableName = "devices")
data class DeviceEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val userId: String? = null,
    val platform: String,
    val name: String,
    val appVersion: String = "",
    val osVersion: String? = null,
    val localOnlyScreenshots: Boolean = true,
    val isCaptureAgent: Boolean = false,
    val lastSeenAt: String? = null,
    val lastSyncRevision: Long = 0,
    val revokedAt: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
)

@Entity(tableName = "user_settings")
data class UserSettingsEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val userId: String,
    val screenshotIntervalMinutes: Double = 5.0,
    val screenshotCaptureEnabled: Boolean = true,
    val pomodoroWorkMinutes: Double = 25.0,
    val pomodoroBreakMinutes: Double = 5.0,
    val localOnlyScreenshots: Boolean = true,
    val retentionDays: Int = 0,
    val idleThresholdMinutes: Double = 5.0,
    val visionAnalysisEnabled: Boolean = false,
    val visionPrivacyRedactionEnabled: Boolean = true,
    val visionModel: String = "",
    val visionConfidenceThreshold: Double = 0.72,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
    val pendingLocal: Boolean = false,
)

@Entity(tableName = "workspace_policy")
data class WorkspacePolicyEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val screenshotUploadAllowed: Boolean = false,
    val syncVisionRawResponse: Boolean = false,
    val defaultLocalOnlyScreenshots: Boolean = true,
    val screenshotRetentionDays: Int = 0,
    val blobRetentionDays: Int = 30,
    val requireApprovalBeforeExport: Boolean = false,
    val idleThresholdMinutes: Double = 5.0,
    val lockedThrough: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val serverRevision: Long = 0,
    val deletedAt: String? = null,
    val originDeviceId: String? = null,
)

// ---------------------------------------------------------------- sync ----

/**
 * The push queue.
 *
 * Ordered (by [seq]), at-least-once, and idempotent on replay because
 * [mutationId] is minted on the client before the first attempt and never
 * changes. A retry after an ambiguous failure re-sends the *same* id, and the
 * server answers `replayed: true` with the original outcome instead of applying
 * the change twice (protocol 7.4).
 *
 * [userId] records the principal the mutation was authored under. A queue built
 * by one user must never be flushed under another's token after an account
 * switch; that is a data-leak bug, not a merge conflict.
 */
@Entity(
    tableName = "pending_mutations",
    indices = [Index(value = ["mutationId"], unique = true), Index(value = ["workspaceId", "seq"])],
)
data class PendingMutationEntity(
    @PrimaryKey(autoGenerate = true) val seq: Long = 0,
    val mutationId: String,
    val workspaceId: String,
    val userId: String?,
    val entity: String,
    val op: String,
    /** The row this mutation targets, so the UI can mark it pending. */
    val targetId: String,
    /**
     * The `server_revision` the edit was made against, or null for a create.
     * The server compares it and answers `stale_write` if the row moved on.
     */
    val baseRevision: Long? = null,
    val payloadJson: String,
    val createdAt: Long,
    val attempts: Int = 0,
    val lastAttemptAt: Long? = null,
    val lastError: String? = null,
    /**
     * Set when the server answered `rejected` with a terminal reason. The row
     * stays in the table so the user can be shown what was dropped and why,
     * rather than the edit evaporating.
     */
    val deadLettered: Boolean = false,
    val deadLetterReason: String? = null,
)

/** One row per workspace. The pull cursor and its bookkeeping. */
@Entity(tableName = "sync_state")
data class SyncStateEntity(
    @PrimaryKey val workspaceId: String,
    /** `server_revision` high-water mark. The next pull asks for `since=cursor`. */
    val cursor: Long = 0,
    val tombstoneHorizonRevision: Long = 0,
    val lastPullAt: Long? = null,
    val lastPushAt: Long? = null,
    val lastSuccessAt: Long? = null,
    val lastError: String? = null,
    /**
     * Set when the server's tombstone horizon has advanced past our cursor: the
     * device slept through the retention window and can no longer learn which
     * rows were deleted. The only correct response is a full resync (protocol
     * 7.6), not a silent continue.
     */
    val resyncRequired: Boolean = false,
)

/**
 * Conflicts and server-raised concerns that a human has to look at.
 *
 * `suspected_duplicate` and `billing_overlap` land here and are NEVER
 * auto-resolved: both mean "these hours may be billed twice", and a client
 * silently picking a winner is how a customer gets invoiced wrong. The device
 * surfaces them and the user decides.
 */
@Entity(tableName = "review_items", indices = [Index(value = ["workspaceId", "resolvedAt"])])
data class ReviewItemEntity(
    @PrimaryKey val id: String,
    val workspaceId: String,
    val entity: String,
    val targetId: String,
    /** `ReviewReasonCodeDto`, or a mutation `reason` for a push conflict. */
    val code: String,
    val detail: String? = null,
    val relatedId: String? = null,
    val raisedAt: String,
    val raisedBy: String = "server",
    val resolution: String = "pending",
    val resolvedAt: String? = null,
    val createdAt: Long,
)

File: projects/timely.noizu.com/apps/backend/lib/timely/sync/entities.ex
defmodule Timely.Sync.Entities do
  @moduledoc """
  The registry that ties an `EntityKind` to its schema, its table, its
  `ChangeSet` bucket, and its wire shape.

  Everything that needs to iterate "all nine buckets" or "the ten entity kinds"
  reads it from here, so adding an entity is one edit rather than nine.

  Two shapes differ between the database and the wire and are reconciled here
  rather than in each caller:

  - `time_span.start_at` / `end_at` are `start` / `end` on the wire. The SQL
    words are reserved; the contract keeps the short names.
  - `settings` stores its body in a `document` jsonb column, but the contract
    flattens that document to top level under a `kind` discriminator. See
    `wire/2` and `payload_to_attrs/2`.
  """

  alias Timely.Schema.Devices.Device
  alias Timely.Schema.Evidence.CensoredScreenshot
  alias Timely.Schema.Evidence.Screenshot
  alias Timely.Schema.Evidence.VisionAnalysis
  alias Timely.Schema.Settings.Setting
  alias Timely.Schema.Taxonomy
  alias Timely.Schema.Tracking.TimeSpan

  # kind => {schema module, changes bucket}
  #
  # `user_settings` and `workspace_policy` are two kinds over one table and one
  # bucket; the union is discriminated by the `kind` column.
  @registry %{
    "client" => {Taxonomy.Client, "clients"},
    "project" => {Taxonomy.Project, "projects"},
    "ticket" => {Taxonomy.Ticket, "tickets"},
    "time_span" => {TimeSpan, "time_spans"},
    "screenshot" => {Screenshot, "screenshots"},
    "vision_analysis" => {VisionAnalysis, "vision_analyses"},
    "censored_screenshot" => {CensoredScreenshot, "censored_screenshots"},
    "device" => {Device, "devices"},
    "user_settings" => {Setting, "settings"},
    "workspace_policy" => {Setting, "settings"}
  }

  # The nine buckets, each with the schema whose rows fill it.
  @buckets [
    {"clients", Taxonomy.Client, "client"},
    {"projects", Taxonomy.Project, "project"},
    {"tickets", Taxonomy.Ticket, "ticket"},
    {"time_spans", TimeSpan, "time_span"},
    {"screenshots", Screenshot, "screenshot"},
    {"vision_analyses", VisionAnalysis, "vision_analysis"},
    {"censored_screenshots", CensoredScreenshot, "censored_screenshot"},
    {"devices", Device, "device"},
    {"settings", Setting, "settings"}
  ]

  @envelope_wire_fields [
    :id,
    :workspace_id,
    :created_at,
    :updated_at,
    :server_revision,
    :deleted_at,
    :origin_device_id
  ]

  # Append-only entities: conflict matrix rows 14 and 15 reject every `update`.
  @append_only ~w(vision_analysis censored_screenshot)

  # Name-keyed entities, whose ids are derived rather than random (3.2) and
  # whose creates take the `duplicate_name` path (rows 12 and 13).
  @taxonomy ~w(client project ticket)

  @doc "All nine `ChangeSet` buckets as `{bucket, schema, kind}`."
  # ⟦𓃀𓎡𓏏𓋴⟧ buckets :: The nine ChangeSet buckets.
  def buckets, do: @buckets

  @doc "The ten `EntityKind` values."
  # ⟦𓎡𓈖𓂧𓋴⟧ kinds :: The ten EntityKind values.
  def kinds, do: Map.keys(@registry)

  @doc "The schema module backing an entity kind."
  # ⟦𓋴𓎡𓅓𓂧⟧ schema :: Schema module for an entity kind.
  def schema(kind) do
    case Map.fetch(@registry, kind) do
      {:ok, {module, _bucket}} -> {:ok, module}
      :error -> {:error, :unknown_entity}
    end
  end

  @doc "The `ChangeSet` bucket an entity kind's rows appear in."
  # ⟦𓃀𓅓𓎡𓏏⟧ bucket :: ChangeSet bucket for an entity kind.
  def bucket(kind) do
    {_module, bucket} = Map.fetch!(@registry, kind)
    bucket
  end

  @doc "True for entities whose every `update` is rejected as `immutable_entity`."
  # ⟦𓄿𓊪𓊪𓈖⟧ append_only? :: True for append-only entity kinds.
  def append_only?(kind), do: kind in @append_only

  @doc "True for the name-keyed `client` / `project` / `ticket` kinds."
  # ⟦𓏏𓐍𓈖𓋴⟧ taxonomy? :: True for name-keyed taxonomy kinds.
  def taxonomy?(kind), do: kind in @taxonomy

  @doc """
  Renders a row in the shape the OpenAPI contract defines.

  `opts[:blob_url_builder]` is a 1-arity function used to render
  `screenshot.blob_url`; omitted, the field is null.
  """
  # ⟦𓅱𓇋𓂋𓋴⟧ wire :: Renders a row in the contract's wire shape.
  def wire(row, opts \\ [])

  def wire(%Taxonomy.Client{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [:name, :canonical_name, :notes, :auto_created, :merged_into_id, :review_state])
    )
  end

  def wire(%Taxonomy.Project{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :name,
        :canonical_name,
        :client_id,
        :client_name,
        :notes,
        :auto_created,
        :merged_into_id,
        :review_state
      ])
    )
  end

  def wire(%Taxonomy.Ticket{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :name,
        :canonical_name,
        :client_id,
        :project_id,
        :client_name,
        :project_name,
        :notes,
        :auto_created,
        :merged_into_id,
        :review_state
      ])
    )
  end

  def wire(%TimeSpan{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :title,
        :client_id,
        :project_id,
        :ticket_id,
        :client_name,
        :project_name,
        :ticket_name,
        :source,
        :is_billable,
        :notes,
        :review_state,
        :derived_from_span_ids
      ])
    )
    |> Map.merge(%{
      # The SQL columns are start_at / end_at because `start` and `end` are
      # reserved words; the contract keeps the short names.
      "start" => iso(row.start_at),
      "end" => iso(row.end_at),
      "locked_at" => iso(row.locked_at),
      "updated_at_effective" => iso(row.updated_at_effective),
      "review_reasons" => row.review_reasons || []
    })
  end

  def wire(%Screenshot{} = row, opts) do
    builder = Keyword.get(opts, :blob_url_builder)

    row
    |> envelope()
    |> Map.merge(take(row, [:span_id, :file_name, :active_app_name, :upload_state]))
    |> Map.merge(%{
      "captured_at" => iso(row.captured_at),
      # Derived, never stored: "are the bytes here right now".
      "blob_available" => row.upload_state == "uploaded",
      "blob_content_hash" => row.blob_content_hash,
      "blob_byte_size" => row.blob_byte_size,
      "blob_uploaded_at" => iso(row.blob_uploaded_at),
      "blob_url" =>
        if(row.upload_state == "uploaded" and is_function(builder, 1),
          do: builder.(row.id),
          else: nil
        )
    })
  end

  def wire(%VisionAnalysis{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :screenshot_id,
        :model,
        :status_update,
        :inferred_project,
        :inferred_task,
        :project_switch_detected,
        :confidence,
        :evidence,
        :privacy_sensitive,
        :privacy_category,
        :raw_response,
        :raw_response_withheld,
        :error_message
      ])
    )
    |> Map.put("analyzed_at", iso(row.analyzed_at))
  end

  def wire(%CensoredScreenshot{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :screenshot_id,
        :span_id,
        :file_name,
        :active_app_name,
        :model,
        :category,
        :reason,
        :confidence,
        :deleted_local_file
      ])
    )
    |> Map.merge(%{
      "captured_at" => iso(row.captured_at),
      "censored_at" => iso(row.censored_at)
    })
  end

  def wire(%Device{} = row, _opts) do
    row
    |> envelope()
    |> Map.merge(
      take(row, [
        :user_id,
        :platform,
        :name,
        :app_version,
        :os_version,
        :local_only_screenshots,
        :is_capture_agent,
        :last_sync_revision
      ])
    )
    |> Map.merge(%{
      "last_seen_at" => iso(row.last_seen_at),
      "revoked_at" => iso(row.revoked_at)
    })
  end

  # The stored `document` is flattened to top level under the `kind`
  # discriminator, which is the shape `SettingsRow` describes. Defaults are
  # merged underneath so a row written before a field existed still reads as a
  # complete document.
  def wire(%Setting{} = row, _opts) do
    defaults =
      case row.kind do
        "workspace_policy" -> Setting.policy_defaults()
        _ -> Setting.user_defaults()
      end

    base =
      row
      |> envelope()
      |> Map.put("kind", row.kind)

    base =
      if row.kind == "user_settings",
        do: Map.put(base, "user_id", row.user_id),
        else: base

    defaults
    |> Map.merge(row.document || %{})
    |> Map.merge(base)
  end

  @doc """
  Folds a wire payload into schema attributes: renames `start` / `end`, and for
  settings gathers every non-envelope key back into `document`.
  """
  # ⟦𓊪𓄿𓏏𓂋⟧ payload_to_attrs :: Wire payload to schema attributes.
  def payload_to_attrs("time_span", payload) do
    payload
    |> Map.drop(["start", "end"])
    |> maybe_put("start_at", payload, "start")
    |> maybe_put("end_at", payload, "end")
  end

  def payload_to_attrs(kind, payload) when kind in ["user_settings", "workspace_policy"] do
    reserved =
      ~w(id workspace_id created_at updated_at updated_at_effective server_revision
         deleted_at origin_device_id kind user_id)

    document = Map.drop(payload, reserved)

    payload
    |> Map.take(reserved)
    |> Map.put("kind", kind)
    |> Map.put("document", document)
  end

  def payload_to_attrs(_kind, payload), do: payload

  # `end: null` must be distinguishable from `end` absent - the first reopens a
  # span (rows 6 and 7), the second leaves it alone - so the key is only carried
  # across when the client actually sent it.
  defp maybe_put(attrs, target, payload, source) do
    if Map.has_key?(payload, source),
      do: Map.put(attrs, target, Map.get(payload, source)),
      else: attrs
  end

  defp envelope(row) do
    @envelope_wire_fields
    |> Enum.map(fn field -> {Atom.to_string(field), encode(Map.get(row, field))} end)
    |> Map.new()
  end

  defp take(row, fields) do
    fields
    |> Enum.map(fn field -> {Atom.to_string(field), encode(Map.get(row, field))} end)
    |> Map.new()
  end

  defp encode(%DateTime{} = value), do: DateTime.to_iso8601(value)
  defp encode(value), do: value

  defp iso(nil), do: nil
  defp iso(%DateTime{} = value), do: DateTime.to_iso8601(value)
  defp iso(value), do: value
end
